packages/grid/x-data-grid/src/hooks/features/columnGrouping/useGridColumnGrouping.ts

```diff
@@ -0,0 +1,183 @@
+import * as React from 'react';
+import { GridApiCommunity } from '../../../models/api/gridApiCommunity';
+import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
+import { GridStateInitializer } from '../../utils/useGridInitializeState';
+import {
+  GridColumnGroupingModel,
+  GridColumnNode,
+  isLeaf,
+} from '../../../models/gridColumnGrouping';
+import { gridColumnGroupsLookupSelector } from './gridColumnGroupsSelector';
+import { gridColumnLookupSelector } from '../columns/gridColumnsSelector';
+import { GridColumnGroupLookup } from './gridColumnGroupsInterfaces';
+import { GridColumnGroupingApi } from '../../../models/api/gridColumnGroupingApi';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridStateColDef, GridColDef } from '../../../models/colDef';
+
+export function hasGroupPath(
+  lookupElement: GridColDef | GridStateColDef,
+): lookupElement is GridStateColDef {
+  return (<GridStateColDef>lookupElement).groupPath !== undefined;
+}
+
+type UnwrappedGroupingModel = { [key: string]: string[] };
+
+// This is the recurrence function that help writing `unwrapGroupingColumnModel()`
+const recurrentUnwrapGroupingColumnModel = (
+  columnGroupNode: GridColumnNode,
+  parents: any,
+): UnwrappedGroupingModel => {
+  if (isLeaf(columnGroupNode)) {
+    return { [columnGroupNode.field]: parents };
+  }
+  const rep: UnwrappedGroupingModel = {};
+  const { groupId, children } = columnGroupNode;
+  children.forEach((child) => {
+    const unwrappedSubTree = recurrentUnwrapGroupingColumnModel(child, [...parents, groupId]);
+    Object.entries(unwrappedSubTree).forEach(([key, value]) => {
+      if (rep[key] !== undefined) {
+        throw new Error(
+          [
+            `MUI: columnGroupingModel contains duplicated field`,
+            `column field ${key} occurrs two times in the grouping model:`,
+            `- ${rep[key].join(' > ')}`,
+            `- ${value.join(' > ')}`,
+          ].join('\n'),
+        );
+      }
+      rep[key] = value;
+    });
+  });
+  return rep;
+};
+
+/**
+ * This is a function that provide for each column the array of its parents.
+ * Parents are ordered from the root to the leaf.
+ * @param columnGroupingModel The model such as provided in DataGrid props
+ * @returns An object `{[field]: groupIds}` where `groupIds` is the parents of the column `field`
+ */
+export const unwrapGroupingColumnModel = (
+  columnGroupingModel?: GridColumnGroupingModel,
+): UnwrappedGroupingModel => {
+  if (!columnGroupingModel) {
+    return {};
+  }
+
+  const rep: UnwrappedGroupingModel = {};
```

Comment:
Maybe rename `rep` to something more meaningful?

```suggestion
  const unwrappedGroupingModel: UnwrappedGroupingModel = {};
```